src/Foundation/NSUrlSessionHandler.cs

```diff
@@ -181,16 +182,20 @@ public NSUrlSessionHandler (NSUrlSessionConfiguration configuration)
 
 		void AddNotification ()
 		{
-			if (!isBackgroundSession && notificationToken == null)
-				notificationToken = NSNotificationCenter.DefaultCenter.AddObserver (UIApplication.WillResignActiveNotification, BackgroundNotificationCb);
+			lock (notificationTokenLock) {
+				if (!isBackgroundSession && notificationToken == null)
+					notificationToken = NSNotificationCenter.DefaultCenter.AddObserver (UIApplication.WillResignActiveNotification, BackgroundNotificationCb);
+			} // lock
 		}
 
 		void RemoveNotification ()
 		{
-			if (notificationToken != null) {
-				NSNotificationCenter.DefaultCenter.RemoveObserver (notificationToken);
-				notificationToken = null;
-			}
+			lock (notificationTokenLock) {
+				if (notificationToken != null) {
+					NSNotificationCenter.DefaultCenter.RemoveObserver (notificationToken);
```

Comment:
I think something like this would be better, to avoid calling RemoveObserver from inside the lock:

```csharp
NSObject localNotificationToken;
lock (notificationTokenLock) {
    localNotificationToken = notificationToken;
    notificationToken = null;
}
if (localNotificationToken != null)
    NSNotificationCenter.DefaultCenter.RemoveObserver (localNotificationToken);
```